File: services/js/history.ts
// One shared timeline for every language service.
const HISTORY_PATH = "../../data/history.json";

export interface HistoryEntry {
    txHash: string;
    message: string;
    lovelace: string;
    to: string;
    sentAt: string;
}

export async function readHistory(): Promise<HistoryEntry[]> {
    const file = Bun.file(HISTORY_PATH);
    if (!(await file.exists())) return [];

    try {
        return await file.json();
    } catch {
        // A corrupted file should not take the server down.
        return [];
    }
}

/** Prepends an entry so the newest transaction is always first. */
export async function addToHistory(entry: HistoryEntry): Promise<void> {
    const history = await readHistory();
    history.unshift(entry);
    await Bun.write(HISTORY_PATH, JSON.stringify(history, null, 2));
}

File: services/go/history.go
package main

import (
	"encoding/json"
	"os"
)

type HistoryEntry struct {
	TxHash   string `json:"txHash"`
	Message  string `json:"message"`
	Lovelace string `json:"lovelace"`
	To       string `json:"to"`
	SentAt   string `json:"sentAt"`
}

func readHistory() ([]HistoryEntry, error) {
	data, err := os.ReadFile("../../data/history.json")
	if os.IsNotExist(err) {
		return []HistoryEntry{}, nil
	}
	if err != nil {
		return nil, err
	}
	var history []HistoryEntry
	if err := json.Unmarshal(data, &history); err != nil {
		return nil, err
	}
	return history, nil
}

func addToHistory(entry HistoryEntry) error {
	history, err := readHistory()
	if err != nil {
		return err
	}
	history = append([]HistoryEntry{entry}, history...)
	data, err := json.MarshalIndent(history, "", "  ")
	return os.WriteFile("../../data/history.json", data, 0644)

}
